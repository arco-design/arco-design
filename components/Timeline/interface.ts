import { CSSProperties, ReactNode } from 'react';

/**
 * @title Timeline
 */
export interface TimelineProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 是否倒序
   * @en Whether Reverse order
   */
  reverse?: boolean;
  /**
   * @zh 时间轴方向
   * @en Timeline direction
   * @defaultValue vertical
   */
  direction?: 'horizontal' | 'vertical';
  /**
   * @zh 时间轴的展示类型：时间轴在左侧/右侧(垂直方向)、上方/下方（水平方向），或交替出现。
   * @en The display mode of Timeline
   * @defaultValue left(vertical) | top(horizontal)
   */
  mode?: 'left' | 'right' | 'top' | 'bottom' | 'alternate' ;
  /**
   * @zh 是否展示幽灵节点，设置为 true 时候只展示幽灵节点。传入ReactNode时，会作为节点内容展示。
   * @en Whether to display ghost nodes. When set to true, only ghost nodes are displayed. When passed to ReactNode, it will be displayed as node content
   */
  pending?: boolean | ReactNode;
  /**
   * @zh 可以传入 ReactNode 定制幽灵节点
   * @en You can pass in ReactNode to customize the ghost node
   * @defaultValue <Spin size={12} />
   */
  pendingDot?: ReactNode;
  /**
   * @zh 设置标签文本的位置
   * @en Position of label text
   * @defaultValue same
   */
  labelPosition?: 'relative' | 'same';
}

/**
 * @title Timeline.Item
 */
export interface TimelineItemProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 节点颜色
   * @en Dot color
   */
  dotColor?: string;
  /**
   * @zh 节点类型：空心圆/实心圆
   * @en Dot type
   * @defaultValue solid
   */
  dotType?: 'hollow' | 'solid';
  /**
   * @zh 自定义节点
   * @en Custom Dot
   */
  dot?: string | ReactNode;
  /**
   * @zh 时间轴类型：实线/虚线/点状线
   * @en Line type
   * @defaultValue solid
   */
  lineType?: 'solid' | 'dashed' | 'dotted';
  /**
   * @zh 时间轴颜色
   * @en Line Color
   */
  lineColor?: string;
  /**
   * @zh 标签文本
   * @en Label text
   */
  label?: string | ReactNode;
  /**
   * @zh 时间轴节点的位置。 在时间轴组件 `mode=alternate` 时候生效
   * @en The position of the timeline node. Take effect when set `mode=alternate`
   */
  labelPosition?: 'relative' | 'same';
  /**
   * @zh 是否自动适配自定义节点尺寸到 16px
   * @en Whether to automatically adapt the custom node size to 16px
   * @defaultValue true
   */
  autoFixDotSize?: boolean;
  position?: 'left' | 'right';
  direction?: 'horizontal' | 'vertical';
  children?: ReactNode;
}
