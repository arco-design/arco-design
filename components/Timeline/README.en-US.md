`````
Component / Data Display

# Timeline

Display information content in chronological or reverse order.
`````

%%Content%%

## API

**Note: Not `Timeline.Item` components will be filtered out and will not be displayed**

### Timeline

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|
|className|Additional css class|`string \| string[]`|`-`|
|reverse|Whether Reverse order|`boolean`|`-`|
|direction|Timeline direction|`'horizontal' \| 'vertical'`|`vertical`|
|mode|The display mode of Timeline|`'left' \| 'right' \| 'top' \| 'bottom' \| 'alternate'`|`left(vertical) \| top(horizontal)`|
|pending|Whether to display ghost nodes. When set to true, only ghost nodes are displayed. When passed to ReactNode, it will be displayed as node content|`boolean \| ReactNode`|`-`|
|pendingDot|You can pass in ReactNode to customize the ghost node|`ReactNode`|`<Spin size={12} />`|
|labelPosition|Position of label text|`'relative' \| 'same'`|`same`|

### Timeline.Item

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|
|className|Additional css class|`string \| string[]`|`-`|
|dotColor|Dot color|`string`|`-`|
|dotType|Dot type|`'hollow' \| 'solid'`|`solid`|
|dot|Custom Dot|`string \| ReactNode`|`-`|
|lineType|Line type|`'solid' \| 'dashed' \| 'dotted'`|`solid`|
|lineColor|Line Color|`string`|`-`|
|label|Label text|`string \| ReactNode`|`-`|
|labelPosition|The position of the timeline node. Take effect when set `mode=alternate`|`'relative' \| 'same'`|`-`|
|autoFixDotSize|Whether to automatically adapt the custom node size to 16px|`boolean`|`true`|
