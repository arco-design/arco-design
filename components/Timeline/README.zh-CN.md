`````
组件 / 数据展示

# 时间轴 Timeline

按照时间顺序或倒序规则的展示信息内容。
`````

%%Content%%

## API

**注意: 非 `Timeline.Item` 组件将会被过滤掉，不会被展示**

### Timeline

|参数名|描述|类型|默认值|
|---|---|---|---|
|style|节点样式|`CSSProperties`|`-`|
|className|节点类名|`string \| string[]`|`-`|
|reverse|是否倒序|`boolean`|`-`|
|direction|时间轴方向|`'horizontal' \| 'vertical'`|`vertical`|
|mode|时间轴的展示类型：时间轴在左侧/右侧(垂直方向)、上方/下方（水平方向），或交替出现。|`'left' \| 'right' \| 'top' \| 'bottom' \| 'alternate'`|`left(vertical) \| top(horizontal)`|
|pending|是否展示幽灵节点，设置为 true 时候只展示幽灵节点。传入ReactNode时，会作为节点内容展示。|`boolean \| ReactNode`|`-`|
|pendingDot|可以传入 ReactNode 定制幽灵节点|`ReactNode`|`<Spin size={12} />`|
|labelPosition|设置标签文本的位置|`'relative' \| 'same'`|`same`|

### Timeline.Item

|参数名|描述|类型|默认值|
|---|---|---|---|
|style|节点样式|`CSSProperties`|`-`|
|className|节点类名|`string \| string[]`|`-`|
|dotColor|节点颜色|`string`|`-`|
|dotType|节点类型：空心圆/实心圆|`'hollow' \| 'solid'`|`solid`|
|dot|自定义节点|`string \| ReactNode`|`-`|
|lineType|时间轴类型：实线/虚线/点状线|`'solid' \| 'dashed' \| 'dotted'`|`solid`|
|lineColor|时间轴颜色|`string`|`-`|
|label|标签文本|`string \| ReactNode`|`-`|
|labelPosition|时间轴节点的位置。 在时间轴组件 `mode=alternate` 时候生效|`'relative' \| 'same'`|`-`|
|autoFixDotSize|是否自动适配自定义节点尺寸到 16px|`boolean`|`true`|
