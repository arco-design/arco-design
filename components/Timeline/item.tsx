import React, { useEffect, useRef, useContext, RefObject } from 'react';
import cs from '../_util/classNames';
import { ConfigContext, ConfigProviderProps } from '../ConfigProvider';
import { TimelineItemProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultProps: TimelineItemProps = {
  dotType: 'solid',
  lineType: 'solid',
};

function Item(baseProps: TimelineItemProps, ref: RefObject<HTMLDivElement>) {
  const { getPrefixCls, componentConfig } = useContext<ConfigProviderProps>(ConfigContext);
  const props = useMergeProps<TimelineItemProps>(
    baseProps,
    defaultProps,
    componentConfig?.['Timeline.Item']
  );

  const {
    children,
    className,
    style,
    label,
    position,
    dot,
    dotColor,
    dotType,
    lineType,
    lineColor,
    direction,
    labelPosition,
    autoFixDotSize = true,
    ...rest
  } = props;

  const dotRef = useRef<HTMLDivElement>();

  const prefixCls = getPrefixCls('timeline');

  const autoFixSize = () => {
    if (dotRef.current) {
      const dotWidth = dotRef.current.offsetWidth;
      const scale = 16 / dotWidth; // 16 是允许的节点最大宽度。
      if (scale < 1) {
        dotRef.current.style.transform = `translateX(-50%) translateY(-50%) scale(${scale})`;
      }
    }
  };

  useEffect(() => {
    autoFixDotSize && autoFixSize();
  }, [props]);

  const labelDom = label && <div className={`${prefixCls}-item-label`}>{label}</div>;

  return (
    <div
      ref={ref}
      role="listitem"
      {...rest}
      className={cs(
        `${prefixCls}-item`,
        {
          [`${prefixCls}-item-${direction}-${position}`]: direction,
          [`${prefixCls}-item-label-${labelPosition}`]: labelPosition,
        },
        className
      )}
      style={style}
    >
      <div className={`${prefixCls}-item-dot-wrapper`}>
        <div
          className={`${prefixCls}-item-dot-line ${prefixCls}-item-dot-line-is-${direction}`}
          style={{
            [`${direction === 'horizontal' ? 'borderTopStyle' : 'borderLeftStyle'}`]: lineType,
            ...(lineColor ? { borderColor: lineColor } : {}),
          }}
        />
        <div className={`${prefixCls}-item-dot-content`}>
          {dot ? (
            <div className={`${prefixCls}-item-dot-custom`} ref={dotRef}>
              {dot}
            </div>
          ) : (
            <div
              className={cs(`${prefixCls}-item-dot`, `${prefixCls}-item-dot-${dotType}`)}
              style={
                dotType === 'solid' ? { backgroundColor: dotColor } : { borderColor: dotColor }
              }
            />
          )}
        </div>
      </div>
      <div className={`${prefixCls}-item-content-wrapper`}>
        {children && <div className={`${prefixCls}-item-content`}>{children}</div>}
        {labelPosition !== 'relative' && labelDom}
      </div>
      {labelPosition === 'relative' && labelDom}
    </div>
  );
}

const ForwardRefItem = React.forwardRef(Item);

const TimelineItem = ForwardRefItem as typeof ForwardRefItem & {
  isTimelineItem?: boolean;
};

TimelineItem.displayName = 'TimelineItem';

TimelineItem.isTimelineItem = true;

export default TimelineItem;

export { TimelineItemProps };
