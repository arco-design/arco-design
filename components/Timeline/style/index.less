@import './token.less';

@timeline-prefix-cls: ~'@{prefix}-timeline';
@timeline-item-prefix-cls: ~'@{prefix}-timeline-item';

.@{timeline-prefix-cls} {
  &-item {
    position: relative;
    color: @timeline-color-content-text;
    font-size: @timeline-font-content-size;
    min-height: @timeline-item-min-height;
    padding-left: @timeline-dot-size-width;

    &-label {
      color: @timeline-color-label-text;
      font-size: @timeline-font-label-size;
      line-height: 1.667;
    }

    &-content {
      line-height: 1.5715;
      font-size: @timeline-font-content-size;
      color: @timeline-color-content-text;
      margin-bottom: @timeline-vertical-margin-content-bottom;

      &-wrapper {
        margin-left: @timeline-vertical-margin-content-left;
        position: relative;
      }
    }
  }
}

// 节点的样式
.@{timeline-prefix-cls}-item {
  &&-last {
    > .@{timeline-prefix-cls}-item-dot-wrapper .@{timeline-prefix-cls}-item-dot-line {
      display: none;
    }
  }

  &-dot-wrapper {
    position: absolute;
    text-align: center;
    height: 100%;
    left: 0;

    .@{timeline-prefix-cls}-item-dot-content {
      width: @timeline-dot-size-width;
      height: 1.5715 * @timeline-font-content-size;
      line-height: $height;
      position: relative;
    }
  }

  &-dot {
    width: @timeline-dot-size-width;
    height: @timeline-dot-size-width;
    border-radius: @timeline-dot-border-radius;
    box-sizing: border-box;
    color: @timeline-color-dot-bg;
    margin-top: -50%;
    position: relative;
    top: 50%;

    &-solid {
      background-color: @timeline-color-dot-bg;
    }

    &-hollow {
      border: @timeline-dot-border-width_hollow solid @timeline-color-dot-bg;
      background-color: @timeline-color-dot-bg_hollow;
    }

    &-custom {
      display: inline-flex;
      transform-origin: center;
      color: @timeline-color-dot-bg;
      box-sizing: border-box;
      position: absolute;
      top: 50%;
      left: 50%;
      background-color: var(~'@{arco-cssvars-prefix}-color-bg-2');
      transform: translateX(-50%) translateY(-50%);

      svg {
        color: inherit;
      }
    }
  }

  &-dot-line {
    width: @timeline-size-line-width;
    border-left-width: @timeline-size-line-width;
    border-color: @timeline-color-line-bg;
    left: 50%;
    transform: translateX(-50%);
    top: 1.5715 * (@timeline-font-content-size / 2) + (@timeline-dot-size-width / 2) +
      @timeline-vertical-margin-line-top;
    bottom: -(1.5715 * (@timeline-font-content-size / 2) - (@timeline-dot-size-width / 2) -
          @timeline-vertical-margin-line-bottom);
    // bottom: -7px;
    position: absolute;
    box-sizing: border-box;
  }
}

/* mode  */
.@{timeline-prefix-cls} {
  &-alternate {
    overflow: hidden;

    .@{timeline-item-prefix-cls}-vertical-left {
      padding-left: 0;

      > .@{timeline-item-prefix-cls}-dot-wrapper {
        left: 50%;
      }

      > .@{timeline-item-prefix-cls}-content-wrapper {
        left: 50%;
        width: 50%;
        margin-left: @timeline-vertical-margin-content-left + @timeline-dot-size-width;
        padding-right: @timeline-vertical-margin-content-left + @timeline-dot-size-width;
      }
    }

    .@{timeline-item-prefix-cls}-vertical-right {
      padding-right: 0;

      > .@{timeline-item-prefix-cls}-dot-wrapper {
        left: 50%;
      }

      > .@{timeline-item-prefix-cls}-content-wrapper {
        left: 0;
        margin-left: -@timeline-vertical-margin-content-left;
        margin-right: 0;
        text-align: right;
        width: 50%;
        padding-right: @timeline-vertical-margin-content-left;
      }
    }
  }

  &-right {
    .@{timeline-item-prefix-cls}-vertical-right {
      padding-right: @timeline-dot-size-width;

      > .@{timeline-item-prefix-cls}-dot-wrapper {
        right: 0;
        left: unset;
      }

      > .@{timeline-item-prefix-cls}-content-wrapper {
        text-align: right;
        margin-left: 0;
        margin-right: @timeline-vertical-margin-content-left;
      }
    }
  }
}

/* label position */
.@{timeline-prefix-cls}-item {
  &-label-relative {
    > .@{timeline-prefix-cls}-item-label {
      position: absolute;
      top: 0;
      max-width: 100px;
      box-sizing: border-box;
    }
  }

  &-vertical-left&-label-relative {
    margin-left: 100px;

    > .@{timeline-prefix-cls}-item-label {
      text-align: right;
      padding-right: @timeline-vertical-margin-content-left;
      transform: translateX(-100%);
      left: 0;
    }
  }

  &-vertical-right&-label-relative {
    margin-right: 100px;

    > .@{timeline-prefix-cls}-item-label {
      text-align: left;
      padding-left: @timeline-vertical-margin-content-left;
      transform: translateX(100%);
      right: 0;
    }
  }

  &-horizontal-top&-label-relative {
    margin-top: 50px;

    > .@{timeline-prefix-cls}-item-label {
      padding-bottom: @timeline-horizontal-margin-content-spacing;
      transform: translateY(-100%);
    }

    > .@{timeline-prefix-cls}-item-content {
      margin-bottom: 0;
    }
  }

  &-horizontal-bottom&-label-relative {
    margin-bottom: 50px;

    > .@{timeline-prefix-cls}-item-content {
      margin-bottom: 0;
    }

    > .@{timeline-prefix-cls}-item-label {
      top: unset;
      bottom: 0;
      text-align: left;
      padding-top: @timeline-horizontal-margin-content-spacing;
      transform: translateY(100%);
    }
  }
}

.@{timeline-prefix-cls}-alternate {
  .@{timeline-prefix-cls}-item-vertical-left.@{timeline-prefix-cls}-item-label-relative {
    margin-left: 0;

    > .@{timeline-prefix-cls}-item-label {
      width: 50%;
      max-width: unset;
      transform: none;
      left: 0;
    }
  }

  .@{timeline-prefix-cls}-item-vertical-right.@{timeline-prefix-cls}-item-label-relative {
    margin-right: 0;

    > .@{timeline-prefix-cls}-item-label {
      width: 50%;
      max-width: unset;
      transform: none;
      right: 0;
    }
  }

  .@{timeline-prefix-cls}-item-horizontal-top.@{timeline-prefix-cls}-item-label-relative {
    margin-top: 0;
  }

  .@{timeline-prefix-cls}-item-horizontal-bottom.@{timeline-prefix-cls}-item-label-relative {
    margin-bottom: 0;
  }
}

@import './horizontal.less';
@import './rtl.less';
