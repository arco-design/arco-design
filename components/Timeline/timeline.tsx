import React, { PropsWithChildren, useContext, ReactElement, RefObject } from 'react';
import cs from '../_util/classNames';
import Item from './item';
import { ConfigContext } from '../ConfigProvider';
import Spin from '../Spin';
import { TimelineProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const getDefaultPosition = (position, index, mode, direction) => {
  let map = ['left', 'right'];
  if (direction === 'horizontal') {
    map = ['top', 'bottom'];
  }
  const res = mode === 'alternate' ? position || map[index % 2] : mode;

  return map.indexOf(res) > -1 ? res : map[0];
};

const defaultProps: TimelineProps = {
  mode: 'left',
  direction: 'vertical',
  pendingDot: <Spin size={12} />,
  labelPosition: 'same',
};

function Timeline(baseProps: PropsWithChildren<TimelineProps>, ref: RefObject<HTMLDivElement>) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  if (rtl) {
    defaultProps.mode = 'right';
  }
  const props = useMergeProps<PropsWithChildren<TimelineProps>>(
    baseProps,
    defaultProps,
    componentConfig?.Timeline
  );
  const {
    className,
    mode,
    reverse,
    children,
    style,
    direction,
    pending,
    labelPosition,
    pendingDot,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('timeline');

  const items =
    React.Children.map(children, (child: ReactElement) => {
      if (child && child.type && (child.type as typeof Item).isTimelineItem) {
        return child;
      }
      return null;
    }) || [];

  if (pending) {
    items.push(
      <Item lineType="dashed" dot={pendingDot}>
        {pending === true ? '' : pending}
      </Item>
    );
  }

  if (reverse) {
    items.reverse();
  }

  return (
    <div
      role="list"
      {...rest}
      ref={ref}
      className={cs(
        prefixCls,
        `${prefixCls}-${mode}`,
        `${prefixCls}-direction-${direction}`,
        { [`${prefixCls}-rtl`]: rtl },
        className
      )}
      style={style}
    >
      {items.map((child: ReactElement, index) => {
        if (!child) {
          return null;
        }

        const { position } = child.props;
        let lineType = child.props.lineType;
        if (pending) {
          if (reverse) {
            if (index === 0) {
              lineType = 'dashed';
            }
          } else if (items.length - 2 === index) {
            lineType = 'dashed';
          }
        }
        return React.cloneElement(child, {
          key: index,
          className: cs(child.props.className, {
            [`${prefixCls}-item-last`]: items.length - 1 === index,
          }),
          labelPosition,
          position: getDefaultPosition(position, index, mode, direction),
          direction,
          lineType,
        });
      })}
    </div>
  );
}

const ForwardRefTimeline = React.forwardRef<unknown, PropsWithChildren<TimelineProps>>(Timeline);

const TimelineComponent = ForwardRefTimeline as typeof ForwardRefTimeline & { Item: typeof Item };

TimelineComponent.displayName = 'Timeline';

TimelineComponent.Item = Item;

export default TimelineComponent;

export { TimelineProps };
