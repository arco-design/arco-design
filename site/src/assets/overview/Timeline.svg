<svg width="128" height="128" viewBox="0 0 128 128" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="47" y="30" width="6" height="6" rx="3" fill="#4080FF"/>
<rect x="57" y="30" width="17" height="2" rx="1" fill="#6AA1FF"/>
<rect x="57" y="34" width="32" height="2" rx="1" fill="#E8F3FF"/>
<path d="M50 40V50" stroke="#94BFFF" stroke-width="2"/>
<path d="M50 64V74" stroke="#94BFFF" stroke-width="2"/>
<path d="M50 88V98" stroke="#94BFFF" stroke-width="2"/>
<rect x="47" y="54" width="6" height="6" rx="3" fill="#4080FF"/>
<rect x="57" y="54" width="17" height="2" rx="1" fill="#6AA1FF"/>
<rect x="57" y="58" width="32" height="2" rx="1" fill="#E8F3FF"/>
<rect x="47" y="78" width="6" height="6" rx="3" fill="#4080FF"/>
<rect x="57" y="77" width="17" height="2" rx="1" fill="#BEDAFF"/>
<rect x="57" y="81" width="32" height="2" rx="1" fill="#E8F3FF"/>
</svg>
