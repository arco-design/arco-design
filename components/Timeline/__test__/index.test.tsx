import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Timeline from '..';

const TimelineItem = Timeline.Item;

mountTest(Timeline);
componentConfigTest(Timeline, 'Timeline');

let wrapper = null;

describe('Timeline', () => {
  beforeEach(() => {
    wrapper && wrapper.unmount();
  });

  it('Prop horizontal & reverse works', () => {
    wrapper = mount(
      <Timeline direction="horizontal" reverse>
        <TimelineItem label="2017-03-10">The first milestone</TimelineItem>
        <TimelineItem label="2018-05-12">The second milestone</TimelineItem>
        <TimelineItem label="2020-09-30">The third milestone</TimelineItem>
      </Timeline>
    );

    expect(
      wrapper
        .find('.arco-timeline')
        .at(0)
        .prop('className')
        .indexOf('arco-timeline-direction-horizontal') > -1
    ).toBe(true);

    expect(wrapper.find('.arco-timeline-item-content').at(0).text()).toBe('The third milestone');
  });

  it('Ignore invalid child', () => {
    wrapper = mount(
      <Timeline>
        <TimelineItem label="2017-03-10">The first milestone</TimelineItem>
        <TimelineItem label="2018-05-12">The second milestone</TimelineItem>
        <TimelineItem label="2020-09-30">The third milestone</TimelineItem>
        <div>hello world</div>
      </Timeline>
    );

    expect(wrapper.find('.arco-timeline-item')).toHaveLength(3);
  });

  it('Pending status', () => {
    wrapper = mount(
      <Timeline pending>
        <TimelineItem label="2017-03-10">The first milestone</TimelineItem>
        <TimelineItem label="2018-05-12">The second milestone</TimelineItem>
        <TimelineItem label="2020-09-30">The third milestone</TimelineItem>
      </Timeline>
    );

    expect(
      wrapper.find('.arco-timeline-item .arco-timeline-item-dot-line').at(2).prop('style')
        .borderLeftStyle
    ).toBe('dashed');
  });
});
