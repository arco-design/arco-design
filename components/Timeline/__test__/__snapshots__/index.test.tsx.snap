// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Timeline default 1`] = `
<div
  class="arco-timeline arco-timeline-left arco-timeline-direction-vertical"
  role="list"
/>
`;

exports[`ConfigProvider componentConfig.Timeline set className globally 1`] = `
<div
  class="arco-timeline arco-timeline-left arco-timeline-direction-vertical global-default-class"
  role="list"
/>
`;

exports[`ConfigProvider componentConfig.Timeline set className globally and component set className itself. 1`] = `
<div
  class="arco-timeline arco-timeline-left arco-timeline-direction-vertical component-class"
  role="list"
/>
`;

exports[`ConfigProvider componentConfig.Timeline.Item default 1`] = `
<div
  class="arco-timeline-item"
  role="listitem"
>
  <div
    class="arco-timeline-item-dot-wrapper"
  >
    <div
      class="arco-timeline-item-dot-line arco-timeline-item-dot-line-is-undefined"
      style="border-left-style:solid"
    />
    <div
      class="arco-timeline-item-dot-content"
    >
      <div
        class="arco-timeline-item-dot arco-timeline-item-dot-solid"
      />
    </div>
  </div>
  <div
    class="arco-timeline-item-content-wrapper"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Timeline.Item set className globally 1`] = `
<div
  class="arco-timeline-item global-default-class"
  role="listitem"
>
  <div
    class="arco-timeline-item-dot-wrapper"
  >
    <div
      class="arco-timeline-item-dot-line arco-timeline-item-dot-line-is-undefined"
      style="border-left-style:solid"
    />
    <div
      class="arco-timeline-item-dot-content"
    >
      <div
        class="arco-timeline-item-dot arco-timeline-item-dot-solid"
      />
    </div>
  </div>
  <div
    class="arco-timeline-item-content-wrapper"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Timeline.Item set className globally and component set className itself. 1`] = `
<div
  class="arco-timeline-item component-class"
  role="listitem"
>
  <div
    class="arco-timeline-item-dot-wrapper"
  >
    <div
      class="arco-timeline-item-dot-line arco-timeline-item-dot-line-is-undefined"
      style="border-left-style:solid"
    />
    <div
      class="arco-timeline-item-dot-content"
    >
      <div
        class="arco-timeline-item-dot arco-timeline-item-dot-solid"
      />
    </div>
  </div>
  <div
    class="arco-timeline-item-content-wrapper"
  />
</div>
`;
